[JIT] Write out SourceRanges on save()

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #22180 [JIT] Load original SourceRanges on import
* **#22179 [JIT] Write out SourceRanges on save()**
* #22178 [JIT] Serialization methods for SourceRange and Source
* #22177 [JIT] Refactor file:line:col to be less ugly

Differential Revision: [D15981423](https://our.internmc.facebook.com/intern/diff/D15981423)

Commit:
[JIT] Preserve SourceRanges across serialization

Add token stack

wire up stack

i fucked up

something salvageable

[JIT] with debug prints

no more debug prints

pickle stuff

Commit:
Update on "[JIT] Preserve SourceRanges across serialization"

[JIT] Preserve SourceRanges across serialization

Add token stack

wire up stack

i fucked up

something salvageable

[JIT] with debug prints

no more debug prints

pickle stuff

gh-metadata: pytorch pytorch 22179 gh/jamesr66a/8/head

Commit:
Update on "[JIT] Preserve SourceRanges across serialization"

[JIT] Preserve SourceRanges across serialization

Add token stack

wire up stack

i fucked up

something salvageable

[JIT] with debug prints

no more debug prints

pickle stuff

gh-metadata: pytorch pytorch 22179 gh/jamesr66a/8/head

Commit:
Update on "[JIT] Preserve SourceRanges across serialization"

[JIT] Preserve SourceRanges across serialization

Add token stack

wire up stack

i fucked up

something salvageable

[JIT] with debug prints

no more debug prints

pickle stuff

gh-metadata: pytorch pytorch 22179 gh/jamesr66a/8/head

Commit:
Update on "[JIT] Preserve SourceRanges across serialization"

[JIT] Preserve SourceRanges across serialization

Add token stack

wire up stack

i fucked up

something salvageable

[JIT] with debug prints

no more debug prints

pickle stuff

gh-metadata: pytorch pytorch 22179 gh/jamesr66a/8/head

zdevito commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -590,8 +644,12 @@ struct PythonPrintPass {
-  bool isLongLine(const std::string& str) {
-    return str.size() + level * 2 >= 40;
+  bool isLongLine(const std::vector<TaggedToken>& expr) {
+    size_t s = 0;
```
This would be cleaner if we just stored the tagged strings as a TaggedString class rather than a vector. It would be better to encapsulate it here rather than letting its internal implementation leak.

zdevito reviewed (commented):
This looks pretty good. I have a couple comments about encapsulating the tagged streams, and making sure we do not create tons of small strings that should be addressed.

Commit:
Update on "[JIT] Write out SourceRanges on save()"

[JIT] Preserve SourceRanges across serialization

gh-metadata: pytorch pytorch 22179 gh/jamesr66a/8/head

zdevito commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -724,17 +802,17 @@ struct PythonPrintPass {
-        std::stringstream ss;
-        ss << "fork(" << name << ")";
-        printOutputDefinition(node, ss.str());
+        auto ss = std::make_shared<TaggedStringStream>(&source_range_stack_);
```
Why does this need to be a TaggedStream? from its creation to its use, the actual source range is never changed.

zdevito commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -167,7 +167,74 @@ const static std::unordered_set<std::string> reserved_names = {
+
+  class TaggedStringStream {
+   public:
+    TaggedStringStream(const SourceRangeStack* srs) : srs(srs) {}
```
use underscores for member names

zdevito commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -167,7 +167,74 @@ const static std::unordered_set<std::string> reserved_names = {
+    TaggedStringStream(TaggedStringStream&& rhs) = default;
+
+    TaggedStringStream& operator<<(const std::string& s) {
+      ranges_.emplace_back((size_t)oss.tellp(), srs->back());
```
This is hugely inefficient! Most of the time the source range is exactly the same as the previous one, but this takes each individual call to << and makes a new 24 byte object to record its source (before serialization, the tuples are larger). This should check that the ranges_.back() has the same source and if so just no-op.

zdevito commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -167,7 +167,74 @@ const static std::unordered_set<std::string> reserved_names = {
+    }
+
+    template <typename T>
+    TaggedStringStream& operator<<(const T& t) {
```
This is also inefficient. It is allocating a new buffer _every time_ << is called. Can we just record it directly into the stream? Even in the real operator it just writes to the real stream anyway.

zdevito commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -402,26 +469,33 @@ struct PythonPrintPass {
   }
-  void assignValue(Value* v, const std::string& s) {
-    value_names_[v] = s;
+  void assignValue(Value* v, std::shared_ptr<TaggedStringStream> s) {
```
This is a complicated case. There are two possible meanings for useOf -- it is a variable name, or it is a inlined definition. When it is a variable name, then typically we _do not_ want to change the source range. For instance on the line `a + b`, we want to report where the add occurred, _not_ where `a` is defined. But if `a` has the source range for where it is defined, then the entire expression will look like it occurs there.

If the value is an inlined definition, then we _do_ want to change the source range. We should probably handle these cases separately (with variable mappings taking a raw string, and inlined definitions taking a TaggedStringStream) in order to get this right.

jamesr66a replied:
Does the python printer emit infix operators anywhere? I just tried your example, and since we emit a prefix expression for add, it works properly:

```
        class FooTest(torch.jit.ScriptModule):
            @torch.jit.script_method
            def forward(self, x, y):
                return x + y

        ft = FooTest()
        loaded = self.getExportImportCopy(ft)
        _, lineno = inspect.getsourcelines(FooTest)

        loaded(torch.rand(3, 4), torch.rand(30, 40))
```

```
The above operation failed in interpreter, with the following stack trace:
at code/tmpcoo1em5o.py:5:10
op_version_set = 1
def forward(self,
    x: Tensor,
    y: Tensor) -> Tensor:
  return torch.add(x, y, alpha=1)
         ~~~~~~~~~ <--- HERE
Compiled from code at test/test_jit.py:3377:24
            @torch.jit.script_method
            def forward(self, x, y):
                return x + y
                       ~~~~~ <--- HERE
"
```

jamesr66a replied:
I fixed this anyway since it would help reduce the amount of data we have to serialize

zdevito commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -402,26 +469,33 @@ struct PythonPrintPass {
+      auto name_stream =
+          std::make_shared<TaggedStringStream>(&source_range_stack_);
+      (*name_stream) << genUniqueNameFor(v);
+      assignValue(v, std::move(name_stream));
```
Example: this is assigning the wrong source to variables, because it gives them the source of their definition and not where the variable was used.

zdevito commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -1066,7 +1151,10 @@ struct PythonPrintPass {
+      auto arg_stream =
+          std::make_shared<TaggedStringStream>(&source_range_stack_);
+      (*arg_stream) << arg_name;
+      assignValue(*param_it++, arg_stream);
```
Also assigning the wrong range to uses.

zdevito reviewed (commented):
This is getting close -- I think there are still some bugs distinguishing places where variables are used and where they are defined. We should not be switching the source range when a variable is used. I also have some concerns about _how big_ the debug information is because the source ranges are not being de-duplicated. There are a lot of really small strings `" = "` That are 3 bytes but will turn into >32 bytes of serialized data in the current format.

jamesr66a commented on torch/csrc/jit/passes/python_print.h:
```diff
@@ -12,8 +12,19 @@ struct Method;
 struct Module;
 } // namespace script
 
+// A pair of (byte offset, SourceRange) describing a specific segment
```
It is a byte offset. The first entry in a SourceRangeRecords will have offset 0. Its size is described by the difference of offset of the next entry and 0

these offsets are retrieved from ostringstream::tellp(), which gives the offset from the beginning of the stream

jamesr66a commented on torch/csrc/jit/export.cpp:
```diff
@@ -921,6 +927,26 @@ void ScriptModuleSerializer::convertModule(
+    // Write out debug records
+    torch::RecordRef* debug_record =
+        module_def->mutable_torchscript_debug_arena();
+    Pickler p;
```
It's moved out into a class in the next diff in the stack

jamesr66a commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -634,15 +709,17 @@ struct PythonPrintPass {
-  void printOutputDefinition(Node* node, const std::string& str) {
+  void printOutputDefinition(
+      Node* node,
+      const std::shared_ptr<TaggedStringStream>& expr) {
```
This is changed in the subsequent PR in the stack

Commit:
Update on "[JIT] Write out SourceRanges on save()"

[JIT] Preserve SourceRanges across serialization

gh-metadata: pytorch pytorch 22179 gh/jamesr66a/8/head

Commit:
Update on "[JIT] Write out SourceRanges on save()"

[JIT] Preserve SourceRanges across serialization

gh-metadata: pytorch pytorch 22179 gh/jamesr66a/8/head

Commit:
Update on "[JIT] Write out SourceRanges on save()"

[JIT] Preserve SourceRanges across serialization

gh-metadata: pytorch pytorch 22179 gh/jamesr66a/8/head

zdevito reviewed (approved):
ngnt

Commit:
Update on "[JIT] Write out SourceRanges on save()"

[JIT] Preserve SourceRanges across serialization

gh-metadata: pytorch pytorch 22179 gh/jamesr66a/8/head
